src/core/storage/sframe_data/sarray_v2_type_encoding.cpp

```diff
@@ -597,7 +577,601 @@ bool typed_decode(const block_info& info,
+  }
+  CORO_END
+  return decode_bufpos;
+}
+
+typed_decode_stream::typed_decode_stream(const block_info& info,
+                                         char* start, size_t len)
+  :info(info),start(start),len(len), iarc(start, len), generic_deserializer{iarc} {
+    // some basic block properties which will be filled in
+    // number of elements
+    dsize = info.num_elem;
+    num_undefined = 0;
+    iarc >> num_types;
+    // if it is a multiple type block, we don't perform a type decode
+    perform_type_decoding = !(info.flags & MULTIPLE_TYPE_BLOCK);
+
+
+    if (!(info.flags & IS_FLEXIBLE_TYPE)) {
+      logstream(LOG_ERROR) << "Attempting to decode a non-typed block"
+                           << std::endl;
+      ASSERT_TRUE(info.flags & IS_FLEXIBLE_TYPE);
+    }
+
+    if (perform_type_decoding) {
+      if (num_types == 1) {
+        //  one block of contiguous type. 
+        char c;
+        iarc >> c;
+        column_type = (flex_type_enum)c;
+      } else if (num_types == 2) {
+        // two types, but with undefined entries.
+        char c;
+        iarc >> c;
+        column_type = (flex_type_enum)c;
+        // read the bitset and undefine all the flagged entries
+        undefined_bitmap.resize(info.num_elem);
+        undefined_bitmap.clear();
+        iarc.read((char*)undefined_bitmap.array, sizeof(size_t)*undefined_bitmap.arrlen);
+        num_undefined = undefined_bitmap.popcount();
+      } else {
+        logstream(LOG_ERROR) << "Unexpected value for num_types: "
+                             << static_cast<int>(num_types)
+                             << " (expected 0, 1, or 2)" << std::endl;
+        ASSERT_TRUE(false);
+      }
+    }
+    elements_to_decode = dsize - num_undefined;
+  }
+
+typed_decode_stream::~typed_decode_stream() {
+  if (number_decoder) delete number_decoder;
+  if (double_decoder) delete double_decoder;
+  if (double_legacy_decoder) delete double_legacy_decoder;
+  if (string_decoder) delete string_decoder;
+  if (vector_decoder) delete vector_decoder;
+  if (ndvector_decoder) delete ndvector_decoder;
+}
+
+/**
+ * Decodes a collection of flexible_type values. The array must be of 
+ * contiguous type, but permitting undefined values.
+ *
+ * See \ref typed_encode() for details.
+ *
+ * \note The coding does not store the number of values stored. This is
+ * stored in the block_info (block.num_elem)
+ */
+size_t typed_decode_stream::read(const std::pair<flexible_type*, size_t>& decodebuffer,
+                                 size_t skip) {
+  if (skip == 0) {
+    ASSERT_TRUE(decodebuffer.first != nullptr && decodebuffer.second > 0);
+  } else {
+    ASSERT_TRUE(decodebuffer.first == nullptr && decodebuffer.second == 0);
+  }
+  size_t decode_bufpos = 0;
+  if (perform_type_decoding) {
+    if (num_types == 1) {
+      if (decodebuffer.first && column_type == flex_type_enum::UNDEFINED) {
+        for (size_t i = 0;i < decodebuffer.second; ++i) {
```

Comment:
Do you want to create a local variable that shadows a class variable, on this line and line 1040, but not on line 1048?